src/Feature/Layout/Layout.re

```diff
@@ -0,0 +1,407 @@
+include AbstractTree;
+
+[@deriving show({with_path: false})]
+type metadata = {size: float};
+
+[@deriving show({with_path: false})]
+type t('id) = AbstractTree.node('id, metadata);
+
+// DSL
+
+module DSL = {
+  let split = (~size=1., direction, children) => {
+    meta: {
+      size: size,
+    },
+    kind: `Split((direction, children)),
+  };
+  let vsplit = (~size=1., children) => split(~size, `Vertical, children);
+  let hsplit = (~size=1., children) => split(~size, `Horizontal, children);
+  let window = (~size=1., id) => {
+    meta: {
+      size: size,
+    },
+    kind: `Window(id),
+  };
+
+  let withSize = (size, node) => {
+    ...node,
+    meta: {
+      size: size,
+    },
+  };
+};
+
+include DSL;
+
+let empty = vsplit([]);
+
+let singleton = id => vsplit([window(id)]);
+
+/**
+ * addWindow
+ */
+let addWindow = (direction, idToInsert, tree) => {
+  switch (tree.kind) {
+  | `Split(_, []) => window(~size=tree.meta.size, idToInsert)
+
+  | `Split(direction, children) =>
+    split(
+      ~size=tree.meta.size,
+      direction,
+      [window(idToInsert), ...children],
+    )
+
+  | `Window(id) =>
+    split(
+      ~size=tree.meta.size,
+      direction,
+      [window(idToInsert), window(id)],
+    )
+  };
+};
+
+let%test_module "addWindow" =
+  (module
+   {
+     let%test "add vertical split" = {
+       let actual = empty |> addWindow(`Vertical, 1);
+
+       actual == window(1);
+     };
+
+     let%test "parent split changes direction if needed" = {
+       let actual =
+         hsplit([window(2), window(1)]) |> addWindow(`Vertical, 3);
+
+       // TODO: Bug
+       actual == hsplit([window(3), window(2), window(1)]);
+       //hsplit([window(1), vsplit([window(3), window(2)])]);
```

Comment:
This existing bug was uncovered when splitting the tests for `addWindow` and `insertWindow` (which used to be combined in `addWindow` with the `~target` argument) so the same test cases are covered for both. Very unlikely to hit this, but should be fixed regardless.

Reply:
Fixed in https://github.com/onivim/oni2/pull/1833/commits/1563e39cee15dfe503c7637e3d5ab2ea203a2bbb